torch_xla/distributed/fsdp/README.md

```diff
@@ -0,0 +1,99 @@
+```bash
+# consolidate the saved checkpoints via command line tool
+python3 -m torch_xla.distributed.fsdp.consolidate_sharded_ckpts \
+  --ckpt_prefix /path/to/your_sharded_checkpoint_files \
+  --ckpt_suffix "_rank-*-of-*.pth"
+```
+
+The implementation of this class is largely inspired by and mostly follows the structure of `fairscale.nn.FullyShardedDataParallel` in https://fairscale.readthedocs.io/en/stable/api/nn/fsdp.html. One of the biggest difference from `fairscale.nn.FullyShardedDataParallel` is that in XLA we don't have explicit parameter storage, so here we resort to a different approach to free full parameters for ZeRO-3.
+
+---
+
+### Example training scripts on MNIST and ImageNet
+
+* MNIST: [`test/test_train_mp_mnist_fsdp_with_ckpt.py`](https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist_fsdp_with_ckpt.py) (it also tests checkpoint consolidation)
+* ImageNet: [`test/test_train_mp_imagenet_fsdp.py`](https://github.com/pytorch/xla/blob/master/test/test_train_mp_imagenet_fsdp.py)
+
+#### Installation
+FSDP is availiabe on PyTorch/XLA 1.12 release and newer nightly. Please refer to https://github.com/pytorch/xla#-available-images-and-wheels for installtion guide.
+
+#### Clone PyTorch/XLA repo
+```bash
+git clone --recursive https://github.com/pytorch/pytorch
+cd pytorch/
+git clone --recursive https://github.com/pytorch/xla.git
+cd ~/
+```
+#### Train MNIST on v3-8 TPU
+It gets around 98.9 accuracy for 2 epochs:
+```bash
+python3 ~/pytorch/xla/test/test_train_mp_mnist_fsdp_with_ckpt.py \
+  --batch_size 16 --drop_last --num_epochs 2 \
+  --use_nested_fsdp --use_gradient_checkpointing
+```
+This script automatically tests checkpoint consolidation at the end. You can also manually consolidate the sharded checkpoints via
+```bash
+# consolidate the saved checkpoints via command line tool
+python3 -m torch_xla.distributed.fsdp.consolidate_sharded_ckpts \
+  --ckpt_prefix /tmp/mnist-fsdp/final_ckpt \
+  --ckpt_suffix "_rank-*-of-*.pth"
+```
+
+#### Train ImageNet with ResNet-50 on v3-8 TPU
+
+It gets around 75.9 accuracy for 100 epochs; download [ImageNet-1k](https://github.com/pytorch/examples/tree/master/imagenet#requirements) to `/datasets/imagenet-1k`:
+```bash
+python3 ~/pytorch/xla/test/test_train_mp_imagenet_fsdp.py \
+  --datadir /datasets/imagenet-1k --drop_last \
+  --model resnet50 --test_set_batch_size 64 --eval_interval 10 \
+  --lr 0.4 --batch_size 128 --num_warmup_epochs 5 --lr_scheduler_divide_every_n_epochs 30 --lr_scheduler_divisor 10 --num_epochs 100 \
+  --use_nested_fsdp
+```
+You can also add ` --use_gradient_checkpointing` (which needs to be used along with `--use_nested_fsdp`) to apply gradient checkpointing on the residual blocks.
+
+---
+
+### Example training scripts on TPU pod (with 10 billion parameters)
+
+To train large models that cannot fit into a single TPU, one should use nested FSDP (wrapping sub-modules with inner FSDP when building the entire model) to implement the ZeRO-3 algorithm.
```

Comment:
Is there a reason we don't have perf numbers for ViT missing from this statement? I see we share it fro ResNet.

Reply:
it was not on ronghang's pr comment. I am just copy that over. Not sure if he has perf number.

Reply:
I guess that it's OK since it focuses on the usage and description of the feature/API?

Reply:
yeah, this is mostly a nit and good to have. we can go ahead without it.